### PR TITLE
Add tryhttp

### PR DESCRIPTION
**Pull request by arp242**
I only implemented the goroutine scheduler for now, which will be
sufficient for getdrip and mailchecker, at least for now. I will add a
RabbitMQ or Redis backend later.

### CODE REVIEW

File: tryhttp.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+	"github.com/teamwork/utils/stringutil"
+)
+
+// Client to perform HTTP requests.
+type Client struct {
+	// Retry determines if we should retry the request after failure. It will
+	// keep retrying until the second argument returns false. The first return
+	// argument can be used to delay the HTTP request.
+	Retry func(r *http.Request, attempt int) (delay time.Duration, retry bool)
```
Comment (ready4god2513) [inline tryhttp.go]: You should specify what we consider a failure.  Is it by http status code?  Something in the body (probably not, but still be good to clarify).

File: tryhttp.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+	// Success callback. This will be run after the HTTP request is finished
+	// without errors or a non-2xx status code.
```
Comment (ready4god2513) [inline tryhttp.go]: Ah, there.  So this is inverse of the `Retry`.  That works.

File: tryhttp.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	if !retry {
+		return
+	}
+
+	c.Scheduler(c, r, attempt, delay, err)
+}
+
+// Goroutines tracks the number of currently active goroutines started with
+// ScheduleGoroutine().
+var Goroutines sync.WaitGroup
```
Comment (ready4god2513) [inline tryhttp.go:110]: Does this need to be public?

File: tryhttp.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		return
+	}
+
+	if resp.Body != nil {
+		resp.Body.Close() // nolint: errcheck
+	}
+
+	delay, retry := c.Retry(r, attempt)
+	if !retry {
+		return
```
Comment (ready4god2513) [inline tryhttp.go]: So there is then no way to get the error back out?

File: tryhttp.go (hunk 3 above)
Comment (arp242) [reply tryhttp.go:110]: The reason I added it is so you have the option to do a "graceful shutdown" by waiting for all goroutines to finish.

So yes :-)

**Comment by cesartw** [comment]
For retry, I used to use https://github.com/facebookgo/httpcontrol
I know this is doing a bit more work but the approach(replacing Transport) it's a lot cleaner

File: tryhttp.go (hunk 4 above)
Comment (arp242) [reply tryhttp.go]: What do you mean?

Comment (ready4god2513) [reply tryhttp.go]: Well, we don't return the response back to the caller anywhere, do we?

**Comment by ready4god2513** [comment]
What happens in case of a server restart @cesartw ?

File: tryhttp.go (hunk 4 above)
Comment (arp242) [reply tryhttp.go]: Ehm, only in case of success. Could add it to the `Retry()` maybe.

**Comment by cesartw** [comment]
I'm not suggesting to use that library, I'm advocating for implementing a Transport so it's less intrusive

File: tryhttp.go (hunk 4 above)
Comment (ready4god2513) [reply tryhttp.go]: I actually like what the facebook package does there.  You can bind a func to a stats type which give you the full request/response back.

**Comment by ready4god2513** [comment]
Ah, yes.  I'd agree about that.

**Comment by arp242** [comment]
I'm not sure that's better. Certainly when serializing stuff to RabbitMQ to be retried hours later (possibly on another server) it's a lot more complex whereas this code is very simple and straightforward.

**Comment by ready4god2513** [comment]
Well, there's a tradeoff between handling complexity in this package and having to handle it in calling packages.  I don't think implementing the RoundTripper and handling serialization is that complicated.

**Comment by cesartw** [comment]
Agree, it will be more complex to implement this library, once. I think it would be easier for ppl using it, tho.

**Comment by arp242** [comment]
?

It's not complex to implement this at all?

**Comment by cesartw** [comment]
I mean, using Transport would make it more complex to implement

**Pull request merged**